chronos/transports/stream.nim

```diff
@@ -1251,171 +1258,91 @@ else:
       return
 
     if WriteClosed in transp.state:
-      transp.state.incl({WritePaused})
-      let error = getTransportUseClosedError()
-      failPendingWriteQueue(transp.queue, error)
-    else:
-      if len(transp.queue) > 0:
-        var vector = transp.queue.popFirst()
-        while true:
-          if transp.kind == TransportKind.Socket:
-            if vector.kind == VectorKind.DataBuffer:
-              let res = posix.send(fd, vector.buf, vector.buflen, MSG_NOSIGNAL)
-              if res >= 0:
-                if vector.buflen - res == 0:
-                  if not(vector.writer.finished()):
-                    vector.writer.complete(vector.size)
-                else:
-                  vector.shiftVectorBuffer(res)
-                  transp.queue.addFirst(vector)
-              else:
-                let err = osLastError()
-                if int(err) == EINTR:
-                  continue
-                else:
-                  try:
-                    transp.fd.removeWriter()
-                  except IOSelectorsException as exc:
-                    raiseAsDefect exc, "removeWriter"
-                  except ValueError as exc:
-                    raiseAsDefect exc, "removeWriter"
+      if transp.queue.len > 0:
```

Comment:
`transp.state.incl({WritePaused})` missing here.

Reply:
Please restore all the `WritePaused` flags everywhere, because it helps a lot with debugging when you can easily understand is writing loop is running or not. Your check with `transp.queue` feels flaky (i'm not sure is it possible to race there but it smells).

Reply:
In the new logic for `writeStreamLoop`, there is a direct relationship between the two things: every time there is something in the queue, `writeStreamLoop` needs to be called to remove it from the queue - the only way `writeStreamLoop` gets called is when you register the socket for event notification and the only time you should register the socket for write notification is when the queue is empty. There's an important interplay with the other flags here as well: `writeClosed` is needed so that nothing is _added_ to the queue after `close`, because there is no longer any way to create the necessary notification, because the socket has been closed.

It was the removal of the flag that led me to finding the bug with missing deregistration from the writer: when there are two ways to express the same thing ("I need a write notification"), subtle bugs like this happen easily.

I'm always wary of having two ways to express a concept or idea because during refactoring or optimization, they tend to go out of sync, as happened with `removeWriter` - I can add it back as a debugging tool, but then, I think we should add a `doAssert` that checks the flag vs the queue length as an invariant - any other "ordering" of operations is a bug. Thoughts?

Reply:
`writePause` has been readded in `resumeWrite`/`removeWriter` along with a `doAssert` to ensure internal consistency with respect to the queue length: 611a8230e25dc04a4633fac59821ee5a3675d394